ignition_ros2_control/src/ignition_system.cpp

```diff
@@ -0,0 +1,402 @@
+IgnitionSystem::export_state_interfaces()
+{
+  std::vector<hardware_interface::StateInterface> state_interfaces;
+
+  for (unsigned int i = 0; i < this->dataPtr->joint_names_.size(); i++) {
+    state_interfaces.emplace_back(
+      hardware_interface::StateInterface(
+        this->dataPtr->joint_names_[i],
+        hardware_interface::HW_IF_POSITION,
+        &this->dataPtr->joint_position_[i]));
+  }
+  for (unsigned int i = 0; i < this->dataPtr->joint_names_.size(); i++) {
+    state_interfaces.emplace_back(
+      hardware_interface::StateInterface(
+        this->dataPtr->joint_names_[i],
+        hardware_interface::HW_IF_VELOCITY,
+        &this->dataPtr->joint_velocity_[i]));
+  }
+  for (unsigned int i = 0; i < this->dataPtr->joint_names_.size(); i++) {
+    state_interfaces.emplace_back(
+      hardware_interface::StateInterface(
+        this->dataPtr->joint_names_[i],
+        hardware_interface::HW_IF_EFFORT,
+        &this->dataPtr->joint_effort_[i]));
+  }
+  return state_interfaces;
+}
+
+std::vector<hardware_interface::CommandInterface>
+IgnitionSystem::export_command_interfaces()
+{
+  std::vector<hardware_interface::CommandInterface> command_interfaces;
+
+  for (unsigned int i = 0; i < this->dataPtr->joint_names_.size(); i++) {
+    command_interfaces.emplace_back(
+      hardware_interface::CommandInterface(
+        this->dataPtr->joint_names_[i],
+        hardware_interface::HW_IF_POSITION,
+        &this->dataPtr->joint_position_cmd_[i]));
+  }
+  for (unsigned int i = 0; i < this->dataPtr->joint_names_.size(); i++) {
+    command_interfaces.emplace_back(
+      hardware_interface::CommandInterface(
+        this->dataPtr->joint_names_[i],
+        hardware_interface::HW_IF_VELOCITY,
+        &this->dataPtr->joint_velocity_cmd_[i]));
+  }
+  for (unsigned int i = 0; i < this->dataPtr->joint_names_.size(); i++) {
+    command_interfaces.emplace_back(
+      hardware_interface::CommandInterface(
+        this->dataPtr->joint_names_[i],
+        hardware_interface::HW_IF_EFFORT,
+        &this->dataPtr->joint_effort_cmd_[i]));
+  }
+  return command_interfaces;
+}
+
+hardware_interface::return_type IgnitionSystem::start()
+{
+  status_ = hardware_interface::status::STARTED;
+  return hardware_interface::return_type::OK;
+}
+
+hardware_interface::return_type IgnitionSystem::stop()
+{
+  status_ = hardware_interface::status::STOPPED;
+  return hardware_interface::return_type::OK;
+}
+
+hardware_interface::return_type IgnitionSystem::read()
+{
+  for (unsigned int j = 0; j < this->dataPtr->joint_names_.size(); j++) {
+    // Get the joint velocity
+    const auto * jointVelocity =
+      this->dataPtr->ecm->Component<ignition::gazebo::components::JointVelocity>(
+      this->dataPtr->sim_joints_[j]);
+
+    // TODO(ahcorde): Revisit this part
+    // Get the joint force
+    // const auto * jointForce =
+    //   this->dataPtr->ecm->Component<ignition::gazebo::components::JointForce>(
+    //   this->dataPtr->sim_joints_[j]);
```

Comment:
There's a known issue with this component: https://github.com/ignitionrobotics/ign-physics/issues/124